src/core/common/settings.cpp

```diff
@@ -305,6 +305,8 @@ otError Settings::ReadFixedSize(Key aKey, void *aBuffer, uint16_t aExpectedSize)
     SuccessOrExit(error = otPlatSettingsGet(&GetInstance(), aKey, 0, reinterpret_cast<uint8_t *>(aBuffer), &size));
     VerifyOrExit(size >= aExpectedSize, error = OT_ERROR_NOT_FOUND);
 
+    memset(aBuffer, 0, aExpectedSize - size);
+
 exit:
```

Comment:
On return from `otPlatSettingsGet()`, the `size` variable should be updated to give the size of `aKey` in the settings.
And we should get here only if `size >= aExpectedSize` (from the `VerifyOrExit` line above), so I think the entire size of `aBuffer` (which is `aExpectedSize`) should already be filled from `otPlatSettingsGet()` call? .
 Maybe I am missing something?

Reply:
This PR is incomplete. In summary, we need to relax the size constraint to allow for SWU where the new version has appended values to the settings structure. If you agree, I will update this PR.

Reply:
Got it. Sounds reasonable to me.

We can consider adding a new method to handle such a case (i.e., allow the saved value to be smaller than what we expect to read) or we can use/update `Read()` to `memset` the left over in the given buffer and use it instead?. The advantage I see by having a different method is that we can indicate which keys we would allow incomplete saved data and which ones we won't.

Thoughts?

Reply:
How about requiring the structure of `aBuffer` is already correctly initialized? So even the read length is smaller, there will be no uninitialized value or **illegal** values? Zero is not guaranteed to be legal values.

Reply:
@jwhui Thoughts?

Reply:
We could treat zero as "uninitialized". If the field is zero after reading, we can then initialize it to some valid value. Given that the `PACKED` declaration was just merged, treating zero as "uninitialized" would also handle cases where there was padding before `PACKED`. Thoughts?

Reply:
It seems this way the structure's logic will be scattered in the reading method. If we provide a Init() method for each structure and call it before reading from non-volatile, we can keep the read method logic clear and prevent from adding a lot extra initializing task in future. Thoughts?

Reply:
This is in fact what I am implementing. However, the initial implementation will simply memset the structure to 0 for simplicity. In the future, we can revisit if there is a need to initialize to non-zero values.

Reply:
OK, I will initialize the newly added `mVersion` to a proper value in #4325.